README.md
# python-sha3 (0.2 beta)
[![Build Status](https://travis-ci.org/bjornedstrom/python-sha3.svg?branch=master)](https://travis-ci.org/bjornedstrom/python-sha3)
October 18, 2015

This module implements the SHA-3 standard as defined in FIPS202: "SHA-3 Standard:  Permutation-Based Hash and Extendable-Output Functions". More specifically, it implements the constructs:

- SHA3-224
- SHA3-256
- SHA3-384
- SHA3-512
- SHAKE128
- SHAKE256

The module is written as a Python C extension on top of optimized implementation available on the Keccak website. This yields better performance than the pure Python implementation that is available. The code is tested on various versions of Python 2 and 3.

## Please Note

The FIPS202 specification is slightly different from the Keccak version that won the SHA-3 competition a few years ago. Many of the "sha3" or "keccak" libraries out there (including many for Python) are not FIPS202 compatible, so they compute different results.

This module aims to be FIPS202 compatible.

## Usage

### SHA-3

The `SHA3*` functions are written to be as similar as possible to `hashlib`:

Sample usage:

    import sha3
    s = sha3.SHA3512() # also 224, 256, 384, 512
                       # also exposed as the function sha3.sha3_512(...)
    s.update('foo')
    print(s.hexdigest())

Importing the `sha3` module will also add the new modules to `hashlib`.

    >>> import hashlib
    >>> import sha3
    >>> hashlib.sha3_512('foo')
    <sha3.SHA3512 object at 0x7fcd0fcb7590>

### SHAKE

The SHAKE functions are Extendable-Output Functions (XOF:s) which are different from hash functions. The interface is similar to `hashlib`. You can either give the output length (in bits) at initialization time or digest computation time:

Alternative 1:

    sk = sha3.SHAKE128(512) # also SHAKE256
    sk.update('')
    print([sk.digest()])

Alternative 2:

    sk = sha3.SHAKE128() # also SHAKE256
    sk.update('')
    print([sk.squeeze(512)])

## Building

    python setup.py build

This will require a C compiler, as usual, and also the Python
development headers.

Optionally, if you want to build a debian package:

    debuild -d -us -uc

## Testing

The `test/` directory contains a bunch of unit tests. By convention
the runnable unit tests have a name that begins with `test_`, such as
the `test/test_usage.py` suite. You can run all the tests with
nosetests:

    $ nosetests test/
    ...
    Ran 1530 tests in 0.155s
    OK

## Caveats

- This module limits itself to FIPS202 behavior, so none if the advanced Keccak behavior are available at the moment.
- At the moment, this code work on byte strings only (i.e. input and output strings must be a multiple if 8 bits).
- Calling the C module directly is a little bit flaky, so it's recommended you only use the functions and classes available from `import sha3`.
- The current implementation most likely has a bug or two, though the unit test coverage is fairly extensive.

## More Information

Please refer to the Keccak website for more information:

http://keccak.noekeon.org/

## Author

This Python module is copyright Björn Edström 2012, 2015 <be@bjrn.se>

The code is heavily based on the Keccak reference which is available here:

https://github.com/gvanas/KeccakCodePackage


debian/changelog
python-sha3 (0.2) UNRELEASED; urgency=low

  * Update with FIPS202.

 -- Björn Edström <be@bjrn.se>  Sun, 18 Oct 2015 18:00:00 +0200

python-sha3 (0.1) UNRELEASED; urgency=low

  * Initial release.

 -- Björn Edström <be@bjrn.se>  Wed, 03 Oct 2012 20:47:04 +0200


sha3.c
/* SHA3 module -- This module provides an interface to SHA-3, also
   known as Keccak.

   This code is based on sha256module.c in the standard Python 2
   distribution, which has the following header:

   ---
   Additional work performed by:

   Andrew Kuchling (amk@amk.ca)
   Greg Stein (gstein@lyra.org)
   Trevor Perrin (trevp@trevp.net)

   Copyright (C) 2005   Gregory P. Smith (greg@krypto.org)
   Licensed to PSF under a Contributor Agreement.
   ---

   This file, adapter for SHA-3, is written by Björn Edström
   <be@bjrn.se> 2012, 2015.

   It is placed under the same license as the original module:
   Licensed to PSF under a Contributor Agreement.
 */

#define _VERSION "0.2.0"

#include "Python.h"
#include "structmember.h"
#include "src/KeccakHash.h"


struct module_state {
};


#if PY_MAJOR_VERSION >= 3
#define GETSTATE(m) ((struct module_state*)PyModule_GetState(m))
#else
#define GETSTATE(m) (&_state)
static struct module_state _state;
#endif


#define MAX_DIGEST_SIZE 2060


typedef struct {
    PyObject_HEAD
    int hashbitlen;
    int outputlen;
    Keccak_HashInstance state;
} SHAobject;


static void SHAcopy(SHAobject *src, SHAobject *dest)
{
    dest->hashbitlen = src->hashbitlen;
    dest->outputlen = src->outputlen;
    memcpy(&dest->state, &src->state, sizeof(src->state));
}


static PyTypeObject SHA3type;


static SHAobject *
newSHA3object(void)
{
    return (SHAobject *)PyObject_New(SHAobject, &SHA3type);
}


static void
SHA_dealloc(PyObject *ptr)
{
    PyObject_Del(ptr);
}


PyDoc_STRVAR(SHA3_copy__doc__, "Return a copy of the hash object.");

static PyObject *
SHA3_copy(SHAobject *self, PyObject *unused)
{
    SHAobject *newobj = NULL;
    if ((newobj = newSHA3object()) == NULL) {
        return NULL;
    }
    SHAcopy(self, newobj);
    return (PyObject *)newobj;
}


PyDoc_STRVAR(SHA3_digest__doc__,
"Return the digest value as a string of binary data.");

static PyObject *
SHA3_digest(SHAobject *self, PyObject *unused)
{
    unsigned char *digest = malloc(self->outputlen / 8 + 1); //[MAX_DIGEST_SIZE];
    SHAobject temp;

    SHAcopy(self, &temp);

    Keccak_HashFinal(&temp.state, digest);

    if (self->hashbitlen > 512) { // SHAKE
	    Keccak_HashSqueeze(&temp.state, digest, self->outputlen);
    }

    PyObject *ret = PyBytes_FromStringAndSize((const char *)digest, self->outputlen / 8);
    free(digest);
    return ret;
}

PyDoc_STRVAR(SHA3_squeeze__doc__,
"Squeeze the digest and return the digest value as a string of binary data.");

static PyObject *
SHA3_squeeze(SHAobject *self, PyObject *args)
{
    int outputlen;

    if (!PyArg_ParseTuple(args, "i", &outputlen))
        return NULL;

    self->outputlen = outputlen;

    return SHA3_digest(self, NULL);
}

PyDoc_STRVAR(SHA3_update__doc__,
"Update this hash object's state with the provided string.");

static PyObject *
SHA3_update(SHAobject *self, PyObject *args)
{
    unsigned char *cp;
    int len;

    if (!PyArg_ParseTuple(args, "s#:update", &cp, &len))
        return NULL;

    Keccak_HashUpdate(&self->state, cp, len * 8);

    Py_INCREF(Py_None);
    return Py_None;
}


PyDoc_STRVAR(SHA3_init__doc__,
"Init this hash object's state.");

static PyObject *
SHA3_init(SHAobject *self, PyObject *args)
{
    int hashbitlen;
    int outputlen;

    if (!PyArg_ParseTuple(args, "ii", &hashbitlen, &outputlen))
        return NULL;

    self->hashbitlen = hashbitlen;
    self->outputlen = outputlen;

    switch (hashbitlen) {
    case 224:
	    Keccak_HashInitialize_SHA3_224(&self->state);
	    break;
    case 256:
	    Keccak_HashInitialize_SHA3_256(&self->state);
	    break;
    case 384:
	    Keccak_HashInitialize_SHA3_384(&self->state);
	    break;
    case 512:
	    Keccak_HashInitialize_SHA3_512(&self->state);
	    break;

    // HACK: This is pretty ugly :-)
    case 10128:
	    Keccak_HashInitialize_SHAKE128(&self->state);
	    break;
    case 10256:
	    Keccak_HashInitialize_SHAKE256(&self->state);
	    break;
    };

    Py_INCREF(Py_None);
    return Py_None;
}


static PyMethodDef SHA_methods[] = {
    {"copy",      (PyCFunction)SHA3_copy,      METH_NOARGS,  SHA3_copy__doc__},
    {"digest",    (PyCFunction)SHA3_digest,    METH_NOARGS,  SHA3_digest__doc__},
    {"squeeze",   (PyCFunction)SHA3_squeeze,   METH_VARARGS, SHA3_squeeze__doc__},
    {"update",    (PyCFunction)SHA3_update,    METH_VARARGS, SHA3_update__doc__},
    {"init"  ,    (PyCFunction)SHA3_init,      METH_VARARGS, SHA3_init__doc__},
    {NULL,        NULL}         /* sentinel */
};


static PyGetSetDef SHA_getseters[] = {
    {NULL}  /* Sentinel */
};


static PyMemberDef SHA_members[] = {
    {NULL}  /* Sentinel */
};


static PyTypeObject SHA3type = {
    PyVarObject_HEAD_INIT(NULL, 0)
    "_sha3.sha3",   /*tp_name*/
    sizeof(SHAobject),  /*tp_size*/
    0,                  /*tp_itemsize*/
    /* methods */
    SHA_dealloc,        /*tp_dealloc*/
    0,                  /*tp_print*/
    0,                  /*tp_getattr*/
    0,                  /*tp_setattr*/
    0,                  /*tp_compare*/
    0,                  /*tp_repr*/
    0,                  /*tp_as_number*/
    0,                  /*tp_as_sequence*/
    0,                  /*tp_as_mapping*/
    0,                  /*tp_hash*/
    0,                  /*tp_call*/
    0,                  /*tp_str*/
    0,                  /*tp_getattro*/
    0,                  /*tp_setattro*/
    0,                  /*tp_as_buffer*/
    Py_TPFLAGS_DEFAULT, /*tp_flags*/
    0,                  /*tp_doc*/
    0,                  /*tp_traverse*/
    0,                  /*tp_clear*/
    0,                  /*tp_richcompare*/
    0,                  /*tp_weaklistoffset*/
    0,                  /*tp_iter*/
    0,                  /*tp_iternext*/
    SHA_methods,        /* tp_methods */
    SHA_members,        /* tp_members */
    SHA_getseters,      /* tp_getset */
};


PyDoc_STRVAR(SHA3_new__doc__,
"Return a new SHA-3 hash object.");

static PyObject *
SHA3_new(PyObject *self, PyObject *args, PyObject *kwdict)
{
    SHAobject *new;

    if ((new = newSHA3object()) == NULL)
        return NULL;

    if (PyErr_Occurred()) {
        Py_DECREF(new);
        return NULL;
    }

    return (PyObject *)new;
}


static struct PyMethodDef SHA_functions[] = {
    {"sha3", (PyCFunction)SHA3_new, METH_VARARGS|METH_KEYWORDS, SHA3_new__doc__},
    {NULL,      NULL}            /* Sentinel */
};


//PyMODINIT_FUNC
//init_sha3(void)


#if PY_MAJOR_VERSION >= 3

static int sha3_traverse(PyObject *m, visitproc visit, void *arg) {
	//Py_VISIT(GETSTATE(m)->AuthenticationError);
    return 0;
}

static int sha3_clear(PyObject *m) {
//Py_CLEAR(GETSTATE(m)->AuthenticationError);
    return 0;
}


static struct PyModuleDef moduledef = {
        PyModuleDef_HEAD_INIT,
        "_sha3",
        NULL,
        sizeof(struct module_state),
        SHA_functions,
        NULL,
        sha3_traverse,
        sha3_clear,
        NULL
};

#define INITERROR return NULL

PyObject *
PyInit__sha3(void)

#else
#define INITERROR return

void
init_sha3(void)
#endif
{
#if PY_MAJOR_VERSION >= 3
    PyObject *module = PyModule_Create(&moduledef);
#else
    PyObject *module = Py_InitModule3("_sha3", SHA_functions, "Module for FIPS202 SHA3/SHAKE");
#endif

    if (module == NULL)
        INITERROR;
    struct module_state *st = GETSTATE(module);

    Py_TYPE(&SHA3type) = &PyType_Type;
    if (PyType_Ready(&SHA3type) < 0) {
        INITERROR;
    }

    {
#if PY_MAJOR_VERSION < 3
        PyObject *s = PyString_FromString(_VERSION);
#else
        PyObject *s = PyUnicode_FromString(_VERSION);
#endif
        PyObject *dict = PyModule_GetDict(module);
        PyDict_SetItemString(dict, "__version__", s);
        Py_DECREF(s);
    }

#if PY_MAJOR_VERSION >= 3
    return module;
#endif
}
